state/machine.go

```diff
@@ -602,6 +602,48 @@ func (m *Machine) PasswordValid(password string) bool {
 // a HasAssignedUnitsError.  If the machine has containers, Destroy
 // will return HasContainersError.
 func (m *Machine) Destroy() error {
+	if m.IsManager() {
+		units, err := m.Units()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		unitsMap := make(map[string]*Unit)
+		for _, unit := range units {
+			unitsMap[unit.String()] = unit
+		}
+		var evacuablePrincipals []string
+		for _, principalUnit := range m.doc.Principals {
+			unit, ok := unitsMap[principalUnit]
+			if !ok {
+				continue
+			}
+			curl, err := unit.CharmURL()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if strings.HasPrefix(curl.Name, "juju-") {
+				evacuablePrincipals = append(evacuablePrincipals, principalUnit)
+			}
+		}
+		if len(m.doc.Principals) != len(evacuablePrincipals) {
+			return newHasAssignedUnitsError(m.doc.Id, m.doc.Principals)
+		}
+		ops := []txn.Op{
+			{
+				C:  machinesC,
+				Id: m.doc.DocID,
+				Assert: bson.D{
+					{"life", Alive},
+					advanceLifecycleUnitAsserts(evacuablePrincipals),
+				},
+			},
+			newCleanupOp(cleanupEvacuateMachine, m.doc.Id),
```

Comment:
The fact that we are not yet advancing the life to dying means that there's a race condition where other units could get deployed.